### PR TITLE
Main group by att keys pr

### PR DESCRIPTION
**Pull request by Simontwice**
for Sebastian's review

### CODE REVIEW

File: research/conditional/moe_layers/cont_moe_designs/group_by_keys.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    def get_merge_and_emit_weights(self, x):
+        keys_from_prev_layer = self.get_from_forward_pass_cache(
+            "attention_keys", self.block_number, "attention"
+        )
+        # shape of keys_from_prev_layer: (batch_size, sequence_length, n_heads, head_size)
+        keys = torch.flatten(keys_from_prev_layer, start_dim=-2)
+        keys = self.reshape_into_token_groups(keys)
+        merge_logits = misc.einsum("B S c d, d e -> B S e c", keys, self.controller)
+        self.update_cache_for_logging("merge_logits", merge_logits)
+        merge_weights = stable_softmax_temperature(merge_logits, self.temperature)
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/cont_moe_designs/group_by_keys.py:26]: shouldn't dim be -2 ?

Comment (sebastianjaszczur) [reply research/conditional/moe_layers/cont_moe_designs/group_by_keys.py:26]: ah, maybe not; assert shape would be useful xD

File: research/conditional/moe_layers/cont_moe_designs/softmax_over_experts_for_emit.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    def get_merge_and_emit_weights(self, x):
+        merge_logits = misc.einsum("B S g d, d e -> B S e g", x, self.controller)
+        self.update_cache_for_logging("merge_logits", merge_logits)
+        merge_weights = stable_softmax_temperature(merge_logits, self.temperature)
+        self.update_cache_for_logging("merge_weights", merge_weights)
+        emit_weights = stable_softmax_temperature(
+            merge_logits, self.temperature, dim=-2
+        )
+        return merge_weights, emit_weights
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/cont_moe_designs/softmax_over_experts_for_emit.py:22]: to nie jest jakiś najlepszy wariant, do mergowania? intuicyjnie powinien być najlepszy chyba, bo jak soft-moe?

File: research/conditional/moe_layers/cont_moe_designs/uniform_routing.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+
+@dataclasses.dataclass(eq=False, repr=False)
+class ContinuousMoEUniformRouting(ContinuousMoeBaseClass):
+
+    """
+    merge weights are calculated based on attention keys from previous layer. The representation is averaged over the head dimension.
+    """
+
+    def get_merge_and_emit_weights(self, x):
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/cont_moe_designs/uniform_routing.py:17]: looks correct

File: research/conditional/moe_layers/cont_moe_designs/uniform_routing.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        merge_logits = misc.einsum("B S g d, d e -> B S e g", x, self.controller)
+        merge_logits = torch.ones_like(merge_logits)
+        self.update_cache_for_logging("merge_logits", merge_logits)
+        merge_weights = stable_softmax_temperature(merge_logits, self.temperature)
+        self.update_cache_for_logging("merge_weights", merge_weights)
```
Comment (sebastianjaszczur) [inline research/conditional/moe_layers/cont_moe_designs/uniform_routing.py:22]: chociaż, w sumie, czy tu nie powinno być dim=-2 ? bo wiem, że wszędzie jest to samo, ale to ma znaczenie który wymiar jest

Comment (sebastianjaszczur) [reply research/conditional/moe_layers/cont_moe_designs/uniform_routing.py:22]: tzn. chodzi mi, że emit powinien mieć dim=-2

**Review by sebastianjaszczur — COMMENTED** [review]
komentarze niżej

File: research/conditional/moe_layers/cont_moe_designs/softmax_over_experts_for_emit.py (hunk 2 above)
Comment (Simontwice) [reply research/conditional/moe_layers/cont_moe_designs/softmax_over_experts_for_emit.py:22]: podrzuce ekspa zaraz, nie ma roznicy

**Comment by sebastianjaszczur** [comment]
Co robimy z PRką? zamykamy, czy będziemy mergować?

**Comment by sebastianjaszczur** [comment]
@Simontwice zamykamy PRkę?
